### DISCUSSION TITLE
Filter using RefinableStringXX mapped to a multi-value person SharePoint Online column

### DISCUSSION THREAD

**Discussion by JoeAyre** [microsoft-search/pnp-modern-search#3582] [Q&A]
Hi,
I'm not sure if this is an issue or a 'by design' thing, but hoping someone has some suggestions that may help me.
Here's the scenario:

- I have a SharePoint Online document library that has a column called MultiplePeople. The column is a person or group column that allows multiple values.
- I have mapped the corresponding OWS_MultiplePeople crawled property to RefinableString00. 
- I have also mapped the corresponding OWS_Q_USER_MultiplePeople crawled property to RefinableString01.
- I have included RefinableString00 and RefinableString01 in a search results web part.
- I have added RefinableString00 and RefinableString01 as filters in a filters web part, using the 'Check box' template and with 'Show count' enabled.
- I have two documents in the library, one that has the MultiplePeople column tagged with John Doe, and one that has the MultiplePeople column tagged with John Doe AND Jane Doe.

The issue I have is that neither filter looks how I want it to look. 
The filter values using RefinableString00 i.e., mapped to OWS_MultiplePeople looks like this:

- [ ] Doe, Jane; Doe, John (1)
- [ ] Doe, John (1)

As you can see, when multiple people are selected, the names are concatenated. 

The filter values using RefinableString01 i.e., mapped to OWS_Q_USER_MultiplePeople looks like this:

- [ ] Jane.Doe@thedomain.com | Doe, Jane | ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF123456 i:0#.f|membership|Jane.Doe@thedomain.com (1)
- [ ] John.Doe@thedomain.com | Doe, John | ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF123456 i:0#.f|membership|John.Doe@thedomain.com (2)

As you can see, whilst the number of filter values (and their total) is correct, the entire user object is returned.

What I would like is the filter to look like this:

- [ ] Doe, Jane (1)
- [ ] Doe, John (2)

Can anyone suggest how I can get the desired appearance?

Note, I accidentally added this to a closed issue https://github.com/microsoft-search/pnp-modern-search/issues/2096 which has some further information, but nothing that has helped me.

Thanks,
Joe

**Comment by wobba** [comment]
You can create a custom filter template in the web part to manipulate the strings using handlebars helper methods.

**Reply by JoeAyre** [reply]
@wobba - thanks for the suggestion. Do you mean within the configuration of the web part via the UI, or do you mean customising the solution via custom code?

**Reply by JoeAyre** [reply]
Just to add, I've creating a custom filter template and referenced this via the 'Use an external template URL' option. My intention is to use the {{split}} helper to split the string on the | character and use the middle portion. Here's the code I'm using within the div class="filter--values-list" div:

`
<div class="filter--values-list">

				{{#each filter.values}}
					<div title="{{#with (split name '|')}}{{#if this.[1]}}{{this.[1]}}{{else}}{{this.[0]}}{{/if}}{{/with}} {{#if ../showCount}}({{count}}){{/if}}" class="filter--value">
						<pnp-filtercheckbox 
							data-instance-id="{{@root.instanceId}}" 
							data-filter-name="{{../filter.filterName}}" 
							data-name="{{#with (split name '|')}}{{#if this.[1]}}{{this.[1]}}{{else}}{{this.[0]}}{{/if}}{{/with}}" 
							data-value="{{value}}" 
							data-selected="{{selected}}" 
							data-disabled="{{disabled}}" 
							data-count="{{count}}"
							data-is-multi="{{../isMulti}}"
							data-theme-variant="{{JSONstringify @root.theme}}">
						</pnp-filtercheckbox>
						{{#if ../showCount}}
							<span data-ui-test-id="filterCount">({{count}})</span>
						{{/if}}
					</div>
				{{/each}}
									
</div>`

Unfortunately, whilst this does split the string and display just the user name e.g., Doe, John, the 'Apply' button no longer becomes activated when I select some filter values.

**Reply by MirasMNC** [reply]
Hi @JoeAyre, have you found the solution how to handle this case? I have same scenario. Thanks! :)

**Reply by JoeAyre** [reply]
Hi @MirasMNC - unfortunately no, still hoping for some advice on this as to why the 'Apply' button becomes inactive after applying the code.

**Reply by wobba** [reply]
Any errors in the browser debug console when you use the template?

**Reply by kasperbolarsen** [reply]
Stand by..... I am looking into this ( actually for my employer 😊 ) and hope to release an update on the PnP Modern Search layouts repository within the week. 
I hope to be able to add a People refiner later this year but can't promise anything 😉

**Comment by kasperbolarsen** [comment]
I am using a RefinableString114 mapped to OWS_Q_USER_SEARCH_PERSONMULTI
and out of the box gives me this
![image](https://github.com/user-attachments/assets/917e8aa7-a126-42a5-a638-079a32299e40)

in the layout I replace line 104 	data-name="{{name}}"  with
data-name= "{{#with (split name '|')}}{{[1]}}{{/with}}" 

which gives me this
![image](https://github.com/user-attachments/assets/ad86c2bd-2e69-4176-b63e-41c071d05900)

It is better than nothing, but I can't figure out a way to update the display value of the selected filter in the layout, I think it will take some rewriting of the pnp-filtersearchbox

**Reply by MirasMNC** [reply]
I have tested it and there is maybe another issue. When I use more filters in the webpart, all of them will be impacted by change the template. It is possible to create custom template I guess only this case.

**Reply by kasperbolarsen** [reply]
I agree. One way to handle that is the wrap the multi value people property in a clause like this:
(it isn't a pretty way to do it, but it is a custom template anyways and it works ;-) )

`
{{#eq filter.filterName 'RefinableString114'}}
		<pnp-filtercheckbox 
			data-instance-id="{{@root.instanceId}}" 
			data-filter-name="{{filter.filterName}}" 
			data-name= "{{#with (split name '|')}}{{[1]}}{{/with}}" 

			data-value="{{value}}" 
			data-selected="{{selected}}" 
			data-disabled="{{disabled}}" 
			data-count="{{count}}"
			data-is-multi="{{../isMulti}}"
			data-theme-variant="{{JSONstringify @root.theme}}">
		</pnp-filtercheckbox>

	{{else}}
		<pnp-filtercheckbox 
		data-instance-id="{{@root.instanceId}}" 
		data-filter-name="{{filter.filterName}}" 
		data-name="{{name}}" 
		data-value="{{value}}" 
		data-selected="{{selected}}" 
		data-disabled="{{disabled}}" 
		data-count="{{count}}"
		data-is-multi="{{../isMulti}}"
		data-theme-variant="{{JSONstringify @root.theme}}"
		>
		</pnp-filtercheckbox>
	{{/eq}}
`

**Reply by kasperbolarsen** [reply]
added to the PnP Modern Search Layouts repo as https://github.com/microsoft-search/pnp-modern-search-layouts/blob/main/Refiner/MultiValuePeopleRefiner/Multi%20value%20people%20property.md

**Reply by MirasMNC** [reply]
Thanks, for my use case it works well even if it is not perfect. :)

**Reply by kasperbolarsen** [reply]
Would prefect look like this?
![image](https://github.com/user-attachments/assets/3d1d508a-1d0b-4f50-8d31-6ac43102f943)

(both the selected value(s) and the mouse over/hint now shows the "short version/display name"

**Reply by MirasMNC** [reply]
I think it looks perferct! :)

**Reply by lucaremdev** [reply]
Hi @kasperbolarsen,
I don't undertsand if the solution you provide here: https://github.com/microsoft-search/pnp-modern-search-layouts/blob/main/Refiner/MultiValuePeopleRefiner/Multi%20value%20people%20property.md
should work or not as of today, to do what JoeAyre described on the top of this discussion.
I just tried to do it, so I copy/pasted the "Layout source code" you gave in [here](https://github.com/microsoft-search/pnp-modern-search-layouts/blob/main/Refiner/MultiValuePeopleRefiner/Multi%20value%20people%20property.md#layout-source-code), and replaced "RefinableString114" by "RefinableString216" (which is the one I use). 
But when I do this, the filter still looks like the first picture you posted [here](https://github.com/microsoft-search/pnp-modern-search/discussions/3582#discussioncomment-10731708), with the format like

> [Jane.Doe@thedomain.com](mailto:Jane.Doe@thedomain.com) | Doe, Jane | ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF1234567890ABCDEF123456 i:0#.f|membership|[Jane.Doe@thedomain.com](mailto:Jane.Doe@thedomain.com)

Did I do something wrong?

Thank you very much!

**Comment by JoeAyre** [comment]
@kasperbolarsen - So close to it being perfect! If only there was a way to handle the data in the pnp-filtersearchbox so that we get just the name rather than the whole object. I've tried and tried to do this but just can't get it to work. It would be great if someone can come up with a solution to this, as it would mean that a person column with mutliple values would be supported by the filters web part. I feel that we are close, but at the moment, having that really ugly object in the pnp-filtersearchbox is just not going to float for our use case - despite it being functional, the users are (somewhat understandably) fixated on getting it looking 'correct'.

**Reply by kasperbolarsen** [reply]
I have spent a considerable amount of time figuring out how to get the filter value to look nice, but without success. 
I think it will require a brand new filter template so it is impossible to predict when or if it will ever show up 🤷

**Reply by calebmil** [reply]
@kasperbolarsen Is the above-mentioned filter template in the works?  I just recently ran into a similar use case with a multi-value managed metadata field (State) where the different elements of the managed metadata item are being treated as separate filter values.  Ultimately, I wind up with three filter values for each available State term as well as a filter value for the parent term set and parent term group.

![image](https://github.com/user-attachments/assets/753a6c92-8f6f-4ec3-8bf5-7ec97874ad93)

**Reply by kasperbolarsen** [reply]
That doesn't look like the issue I am working on,  it looks more like a mapping issue or stale data

**Reply by kasperbolarsen** [reply]
Perhaps enabling localization?

**Reply by calebmil** [reply]
> That doesn't look like the issue I am working on, it looks more like a mapping issue or stale data

I'm not sure where localization would come in.  While we have a multi-geo tenant, we are only working on this within our NA tenant.  No additional languages, just English.

I've tried changing the mapping for the State column and am waiting for the index to update.  Seems like the SLA is no longer 24 hours for requested index refreshes anymore ...
